doc/man/ejsc.1
.TH EC "1" "March 2014" "ejsc" "User Commands"
.SH NAME
ejsc \- Ejscript compiler.
.SH SYNOPSIS
.B ejsc
[\fI--bind\fR]
[\fI--debug\fR]
[\fI--doc\fR]
[\fI--noout\fR]
[\fI--optimize level\fR]
[\fI--out filename\fR]
[\fI--parse\fR]
[\fI--require 'module ...'\fR]
[\fI--search ejsPath\fR]
[\fI--standard\fR]
[\fI--strict\fR]
[\fI--version\fR]
[\fI--warn level\fR]
files...
.SH DESCRIPTION
The \fBejsc\fR command compiles Ejscript programs to produce Ejscript module files containing byte code.
The module files may then be run via the \fBejs\fR shell command.
.PP
The \fBejsc\fR command is both a compiler and link editor. Inputs can be either Ejscript source files or
Ejscript modules that have come from previous invocations of the \fBejsc\fR command. Outputs will be one or more Ejscript 
module files that contain declarations and byte code. 
.PP
An output module file will be created for each Ejscript module directive
encountered during compilation. A module file will also be created for any global variables or functions 
declared outside a module directive. These global declarations will go into the \fBdefault\fR module that is 
specially reserved for global declarations.  Each module file will be named according to the module directive name, 
but with a \fB.mod\fR extension appended. 
.SH LINKING
If \fBejsc\fR is invoked with the --out switch, all input scripts, modules and any other dependent modules are merged together 
into a single output module. The modules retain their logical naming, but are emitted in a single output module file. When 
that module file is loaded, all the contained modules will be available to the program. This creates not only a convenient 
way to package an entire application as a single file, it also permits many optimizations by merging the entire application 
and its dependent modules into a single module.
.SH BINDING
The \fBejsc\fR compiler will attempt to early-bind all possible variable and function references. Binding means resolving 
references to the underlying storage for properties and functions. Doing this at compile time usually results in much
faster execution at run-time.  
.PP
When using the --out switch, the compiler can early-bind all global variables, functions and 
type references, resulting in a much smaller and faster application. However, you must not subsequently load other 
modules that also have global declarations. Otherwise the Ejscript loader will throw an exception. Consequently, 
the --out switch must only be used to create a complete application including all the application's required modules.
.SH COMPLIANCE
Ejscript is fully compliant with the JavaScript (ECMA-262 3.X) standards, but it also offers a set of 
enhancements and fixes designed to improve the language and correct some longstanding JavaScript design issues. 
.SH OPTIONS
.TP
\fB\--bind\fR
Bind global variable declarations into slot references. This results in faster more compact code. However, only
one module can have its global variables bound.
.TP
\fB\--debug\fR
Generate symbolic debug instructions. This permits symbolic debugging of Ejscript programs and enables exception
stack backtraces to include line number information.
.TP
\fB\--doc\fR
Include documentation strings from input scripts in the output modules. The \fBejsmod\fR command can then generate
HTML documentation using these doc strings. The format of the doc strings resembles that of Javadoc.
.TP
\fB\--merge\fR
Merge all input files and modules together into a single output module. This option acts like a link-editor combining
all inputs together. Useful if you want to distribute your application as a single module file.
.TP
\fB\--optimize level\fR
Set the code optimization level. Level values must be between 0 (least) and 9 (most). Default is 9.
.TP
\fB\--parse\fR
Just parse the source scripts. Don't verify, execute or generate output. Useful to check the script syntax only.
.TP
\fB\--require 'module ...'\fR
List of modules to preload before compiling input files.
.TP
\fB\--search ejsPath\fR
Set the module search path. The module search path is a set of directories that the \fBejsc\fR command will use
when locating and loading Ejscript modules.  The search path will always have some system directories appended 
to the end. These include paths specified via the \fB\ EJSPATH\fR environment variable and key system directories
such as the Ejscript system module directory and the directory containing the \fBejsc\fR command.
.IP
The search path value is similar in format to the system PATH variable format. 
On windows, path segments are separated by ";" and on Linux, Unix, FreeBSD and MAC, the path segments are separated 
by ":" delimiters.
.IP
Given a module named "a.b.c" in a script, \fBejsc\fR will use the following search strategy to locate the module:
.IP
1. Search for a module file named "a.b.c.mod"
.IP
2. Search for a module file named "a/b/c.mod"
.IP
3. Search for a module file named "a.b.c.mod" in the search path
.IP
4. Search for a module file named c.mod in the search path
.TP
\fB\--standard\fR
Run scripts in standard mode. Ejscript supports two parsing modes: strict and standard. Standard mode does not require 
variables be declared and typed before use.
.TP
\fB\--strict\fR
Run scripts in standard mode. Ejscript supports two parsing modes: strict and standard. Strict mode requires that 
all variables be declared and typed.
.TP
\fB\--version\fR
Print the \fBejsc\fR command version and exit.
.TP
\fB\--warn level \fR
Set the compiler warning verbosity level. Level values must be between 0 (least verbose) and 9 (most). Default is 0.
.SH BACKGROUND
Ejscript is an enhanced implementation of the JavaScript language for use in embedded applications such as web servers, 
embedded and mobile devices. It is especially suited for Server-Side JavaScript web applications.
.PP
Ejscript is a dynamic, interpreted, object-oriented scripting language that supports classes, objects, 
exceptions, statements, expressions and a powerful suite of data types.
.SH "REPORTING BUGS"
Report bugs to dev@embedthis.com.
.SH COPYRIGHT
Copyright \(co 2004-2014 Embedthis Software.  Ejscript is a trademark of Embedthis Software.
.br
.SH "SEE ALSO"
ejsgen, ejs, ejsmod


src/cmd/ejsc.c
/* 
    ejsc.c - Ejscript Compiler main program

    This compiler will compile the files given on the command line.

    Copyright (c) All Rights Reserved. See details at the end of the file.
 */

/********************************** Includes **********************************/

#include    "ejsCompiler.h"

#if BIT_PACK_EJSCRIPT

/********************************** Forwards **********************************/

typedef struct App {
    Ejs         *ejs;
    EcCompiler  *compiler;
    MprList     *modules;
} App;

static App *app;

static MprList *expandWild(Ejs *ejs, int argc, char **argv);
static void manageApp(App *app, int flags);
static void require(cchar *name);

/************************************ Code ************************************/

MAIN(ejscMain, int argc, char **argv, char **envp)
{
    Mpr             *mpr;
    MprList         *args;
    Ejs             *ejs;
    EcCompiler      *cp;
    char            *argp, *searchPath, *outputFile, *outputDir, *certFile, *name, *tok, *modules;
    int             nextArg, err, ejsFlags, ecFlags, bind, debug, doc, merge, modver;
    int             warnLevel, noout, parseOnly, tabWidth, optimizeLevel, strict;

    /*
        Initialize the Multithreaded Portable Runtime (MPR)
     */
    mpr = mprCreate(argc, argv, 0);
    app = mprAllocObj(App, manageApp);
    mprAddRoot(app);
    mprAddStandardSignals();

    if (mprStart() < 0) {
        mprError("Cannot start mpr services");
        return EJS_ERR;
    }
    err = 0;
    searchPath = 0;
    strict = 0;
    certFile = 0;
    ecFlags = 0;
    bind = 0;
    debug = 0;
    doc = 0;
    merge = 0;
    modver = 0;
    noout = 0;
    parseOnly = 0;
    tabWidth = 4;
    warnLevel = 1;
    outputFile = 0;
    outputDir = 0;
    optimizeLevel = 9;

    for (nextArg = 1; nextArg < argc; nextArg++) {
        argp = argv[nextArg];
        if (*argp != '-') {
            break;
        }
        if (strcmp(argp, "--bind") == 0) {
            bind = 1;

        } else if (strcmp(argp, "--debug") == 0) {
            debug = 1;

        } else if (strcmp(argp, "--debugger") == 0 || strcmp(argp, "-D") == 0) {
            mprSetDebugMode(1);

        } else if (strcmp(argp, "--dir") == 0) {
            /*
                Set the output directory for modules
             */
            if (nextArg >= argc) {
                err++;
            } else {
                outputDir = argv[++nextArg];
            }

        } else if (strcmp(argp, "--doc") == 0) {
            doc = 1;

        } else if (strcmp(argp, "--log") == 0) {
            /*
                Undocumented logging switch
             */
            if (nextArg >= argc) {
                err++;
            } else {
                mprStartLogging(argv[++nextArg], 0);
                mprSetCmdlineLogging(1);
            }

        } else if (strcmp(argp, "--merge") == 0) {
            merge = 1;

        } else if (strcmp(argp, "--modver") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                modver = ejsParseModuleVersion(argv[++nextArg]);
            }

        } else if (strcmp(argp, "--nobind") == 0) {
            bind = 0;

        } else if (strcmp(argp, "--noout") == 0) {
            noout = 1;

        } else if (strcmp(argp, "--standard") == 0) {
            strict = 0;

        } else if (strcmp(argp, "--strict") == 0) {
            strict = 1;

        } else if (strcmp(argp, "--optimize") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                optimizeLevel = atoi(argv[++nextArg]);
            }

        } else if (strcmp(argp, "--out") == 0) {
            /*
                Create a single output module file containing all modules
             */
            if (nextArg >= argc) {
                err++;
            } else {
                outputFile = argv[++nextArg];
            }

        } else if (strcmp(argp, "--parse") == 0) {
            parseOnly = 1;

        } else if (strcmp(argp, "--search") == 0 || strcmp(argp, "--searchpath") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                searchPath = argv[++nextArg];
            }

        } else if (strcmp(argp, "--require") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                if (app->modules == 0) {
                    app->modules = mprCreateList(-1, 0);
                }
                modules = sclone(argv[++nextArg]);
                /*  Fix for Xcode and Visual Studio */
                if (modules[0] == ' ' || scmp(modules, "null") == 0) {
                    modules[0] = '\0';                    
                }
                name = stok(modules, " \t,", &tok);
                while (name != NULL) {
                    require(name);
                    name = stok(NULL, " \t", &tok);
                }
            }

        } else if (strcmp(argp, "--sign") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                certFile = argv[++nextArg];
            }

#if FUTURE
        } else if (strcmp(argp, "--strip") == 0) {
            strip = 1;
#endif
        } else if (strcmp(argp, "--tabWidth") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                tabWidth = atoi(argv[++nextArg]);
            }

        } else if (strcmp(argp, "--version") == 0 || strcmp(argp, "-V") == 0) {
            mprPrintf("%s\n", BIT_TITLE, EJS_VERSION);
            return 0;

        } else if (strcmp(argp, "--warn") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                warnLevel = atoi(argv[++nextArg]);
            }

        } else if (strcmp(argp, "--web") == 0) {
            if (app->modules == 0) {
                app->modules = mprCreateList(-1, 0);
            }
            require("ejs");
            require("ejs.unix");
            require("ejs.db");
            //  TODO - decouple and remove this
            require("ejs.db.mapper");
            require("ejs.web");

        } else {
            err++;
            break;
        }
    }
    if (noout || merge) {
        bind = 1;
    }
    if (outputFile && noout) {
        mprEprintf("Cannot use --out and --noout\n");
        err++;
    }
    if (argc == nextArg) {
        err++;
    }
    if (err) {
        /*
            Usage Examples:
                ejsc Person.es User.es Customer.es
                ejsc --out group.mod Person.es User.es Customer.es
                ejsc --out group.mod Person.es User.es Customer.es

            NOTE: bind is deliberately not documented and is for internal use only.
         */
        mprEprintf("Usage: %s [options] files...\n"
            "  Ejscript compiler options:\n"
            "  --debug                # Include symbolic debugging information in output\n"
            "  --doc                  # Include documentation strings in output\n"
            "  --dir directory        # Set the output directory for modules (default: \".\")\n"
            "  --merge                # Merge dependent input modules into the output\n"
            "  --modver version       # Set the default module version\n"
            "  --noout                # Do not generate any output\n"
            "  --optimize level       # Set optimization level (0-9)\n"
            "  --out filename         # Name a single output module (default: \"default.mod\")\n"
            "  --parse                # Just parse source. No output\n"
            "  --require 'module ...' # List of required modules to pre-load\n"
            "  --search ejsPath       # Module search path\n"
            "  --standard             # Default compilation mode to standard (default)\n"
            "  --strict               # Default compilation mode to strict\n"
#if FUTURE
            "  --sign certFile        # Sign the module file (not implemented) \n"
            "  --strip                # Strip all symbolic names (Cannot import)\n"
            "  --tabwidth             # Tab width for '^' error reporting\n"
#endif
            "  --version              # Emit the compiler version information\n"
            "  --warn level           # Set the warning message level (0-9)\n\n",
            mpr->name);
        return -1;
    }
    ejsFlags = EJS_FLAG_NO_INIT;
    if (doc) {
        ejsFlags |= EJS_FLAG_DOC;
    }
    if ((ejs = ejsCreateVM(0, 0, ejsFlags)) == 0) {
        return MPR_ERR_MEMORY;
    }
#if UNUSED
    mprRunDispatcher(ejs->dispatcher);
#endif
    if (ejsLoadModules(ejs, searchPath, app->modules) < 0) {
        return MPR_ERR_CANT_READ;
    }
    app->ejs = ejs;
    ecFlags |= (debug) ? EC_FLAGS_DEBUG: 0;
    ecFlags |= (merge) ? EC_FLAGS_MERGE: 0;
    ecFlags |= (bind) ? EC_FLAGS_BIND: 0;
    ecFlags |= (noout) ? EC_FLAGS_NO_OUT: 0;
    ecFlags |= (parseOnly) ? EC_FLAGS_PARSE_ONLY: 0;
    ecFlags |= (doc) ? EC_FLAGS_DOC: 0;

    cp = app->compiler = ecCreateCompiler(ejs, ecFlags);
    if (cp == 0) {
        return MPR_ERR_MEMORY;
    }
    cp->require = app->modules;
    cp->modver = modver;

    ecSetOptimizeLevel(cp, optimizeLevel);
    ecSetWarnLevel(cp, warnLevel);
    ecSetStrictMode(cp, strict);
    ecSetTabWidth(cp, tabWidth);
    ecSetOutputDir(cp, outputDir);
    ecSetOutputFile(cp, outputFile);
    ecSetCertFile(cp, certFile);

    if (nextArg < argc) {
        /*
            Compile the source files supplied on the command line. This will compile in-memory and
            optionally also save to module files.
         */
        if ((args = expandWild(ejs, argc - nextArg, &argv[nextArg])) == 0) {
            err++;
        } else if (ecCompile(cp, args->length, (char**) args->items) < 0) {
            err++;
        }
        if (cp->warningCount > 0 || cp->errorCount > 0) {
            mprRawLog(0, "%s\n", cp->errorMsg);
        }
    }
    if (cp->errorCount > 0) {
        err++;
    }
    app->ejs = 0;
    app->compiler = 0;
    ejsDestroy(ejs);
    mprDestroy(MPR_EXIT_IMMEDIATE);
    return err;
}


static MprList *expandWild(Ejs *ejs, int argc, char **argv)
{
    MprList     *list;
    EjsArray    *files;
    EjsPath     *path, *dir;
    cchar       *arg;
    int         i, j;

    if ((list = mprCreateList(-1, 0)) == 0) {
        return 0;
    }
    for (i = 0; i < argc; i++) {
        if (schr(argv[i], '*')) {
            arg = mprNormalizePath(argv[i]);
            dir = ejsCreatePathFromAsc(ejs, mprGetPathDir(arg));
            path = ejsCreatePathFromAsc(ejs, mprGetPathBase(arg));
            if ((files = ejsGetPathFiles(ejs, dir, 1, (EjsObj**) (void*) &path)) == 0) {
                ejsClearException(ejs);
                mprAddItem(list, sclone(argv[i]));
            } else {
                for (j = 0; j < files->length; j++) {
                    mprAddItem(list, ((EjsPath*) files->data[j])->value);
                }
            }
        } else {
            mprAddItem(list, sclone(argv[i]));
        }
    }
    return list;
}


static void manageApp(App *app, int flags)
{
    if (flags & MPR_MANAGE_MARK) {
        mprMark(app->compiler);
        mprMark(app->ejs);
        mprMark(app->modules);
    }
}


static void require(cchar *name) 
{
    assert(app->modules);
    if (name && *name) {
        mprAddItem(app->modules, sclone(name));
    }
}


#else
int main(int argc, char **argv)
{
    return 0;
}

#endif /* BIT_PACK_EJSCRIPT */
/*
    @copy   default

    Copyright (c) Embedthis Software LLC, 2003-2014. All Rights Reserved.

    This software is distributed under commercial and open source licenses.
    You may use the Embedthis Open Source license or you may acquire a 
    commercial license from Embedthis Software. You agree to be fully bound
    by the terms of either license. Consult the LICENSE.md distributed with
    this software for full details and other copyrights.

    Local variables:
    tab-width: 4
    c-basic-offset: 4
    End:
    vim: sw=4 ts=4 expandtab

    @end
 */
